hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -193,8 +194,16 @@ protected HoodieBackedTableMetadataWriter(StorageConfiguration<?> storageConf,
     ValidationUtils.checkArgument(!initialized || this.metadata != null, "MDT Reader should have been opened post initialization");
   }
 
+  @VisibleForTesting
   List<MetadataPartitionType> getEnabledPartitions(HoodieMetadataConfig metadataConfig, HoodieTableMetaClient metaClient) {
-    return MetadataPartitionType.getEnabledPartitions(metadataConfig, metaClient);
+    if (!metadataConfig.isEnabled()) {
+      return Collections.emptyList();
+    }
+    return MetadataPartitionType.getValidValues().stream()
+        .filter(partitionType -> partitionType.isMetadataPartitionSupported(metaClient)
+            && (partitionType.isMetadataPartitionEnabled(metadataConfig)
+            || partitionType.isMetadataPartitionAvailable(metaClient)))
+        .collect(Collectors.toList());
```

Comment:
- isMetadataPartitionSupported
- isMetadataPartitionEnabled
- isMetadataPartitionAvailable

Looks very confusing from high-level, can we clean them up? and instead of passing around metaClient, can we just passes the table config?

Reply:
> a. isMetadataPartitionSupported -> some partitions are not supported in v6, but only version 8 and above.
b. isMetadataPartitionEnabled -> based on writer properties, we determine which partitions are enabled.
c. isMetadataPartitionAvailable -> For secondary and expression index, even if writer properties does not suggest

These discrepencies are so confusing and does not deserve a new API.